src/com/facebook/buck/swift/SwiftDescriptions.java

```diff
@@ -69,4 +60,27 @@ private SwiftDescriptions() {
     output.enableObjcInterop = Optional.of(true);
   }
 
+  static boolean hasSwiftSource(ImmutableSortedSet<SourceWithFlags> srcs) {
+    return Iterables.any(srcs, new Predicate<SourceWithFlags>() {
+      @Override
+      public boolean apply(SourceWithFlags input) {
+        SourcePath sourcePath = input.getSourcePath();
+        return (sourcePath instanceof PathSourcePath) &&
```

Comment:
The previous approach has 2 issues:
1. Since there are extra `swift-lib` targets even when there is no swift source, project generation task will have to filter them out, this is consider overhead. Indeed, project generation has nothing to do with the companion target, we should better off generate xcode project without creating the companion target.
2. There is the case that `AppleTestDescription` delegate its `findDepsForTargetFromConstructorArgs` to corresponding `AppleLibraryDescription`, here is where the mess gets bigger, since we have a companion target for `apple_test` as well, this is make the project generation way more complicated.

Since I haven't started with project generation, I would not touch before I have a clear picture of this part. I will update PR to cover the case of `BuildTargetSourcePath` and will update this when we start to work with project generation.
The change in code would not be big though, sound good to you @Coneko ?